dtfs-central-api/src/repositories/tfm-deals-repo/tfm-deal-cancellation.repo.submitDealCancellation.test.ts

```diff
@@ -0,0 +1,113 @@
+
+const auditDetails = generateTfmAuditDetails(tfmUserId);
+
+describe('tfm-deals-cancellation-repo', () => {
+  const updateOneMock = jest.fn();
+  const getCollectionMock = jest.fn();
+
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  describe('submitDealCancellation', () => {
+    const mockUpdateResult = { matchedCount: 1 };
+
+    beforeEach(() => {
+      updateOneMock.mockResolvedValue(mockUpdateResult);
+
+      getCollectionMock.mockResolvedValue({
+        updateOne: updateOneMock,
+      });
+      jest.spyOn(db, 'getCollection').mockImplementation(getCollectionMock);
+    });
+
+    it('calls the DB with the correct collection name', async () => {
+      // Act
+      await TfmDealCancellationRepo.submitDealCancellation(dealId, mockDealCancellationObject, auditDetails);
+
+      // Assert
+      expect(getCollectionMock).toHaveBeenCalledWith(MONGO_DB_COLLECTIONS.TFM_DEALS);
+      expect(updateOneMock).toHaveBeenCalledTimes(1);
+    });
+
+    it('throws an InvalidDealIdError if deal is not a valid object id', async () => {
+      // Arrange
+      const invalidDealId = 'xyz';
+
+      // Assert
+      await expect(TfmDealCancellationRepo.submitDealCancellation(invalidDealId, mockDealCancellationObject, auditDetails)).rejects.toThrow(
+        new InvalidDealIdError(invalidDealId.toString()),
+      );
+    });
+
+    it('throws a DealNotFoundError if no matching result is found', async () => {
+      // Arrange
+      const mockFailedUpdateResult = { matchedCount: 0 };
+
+      updateOneMock.mockResolvedValue(mockFailedUpdateResult);
+
+      getCollectionMock.mockResolvedValue({ updateOne: updateOneMock });
+
+      // Assert
+      await expect(TfmDealCancellationRepo.submitDealCancellation(dealId, mockDealCancellationObject, auditDetails)).rejects.toThrow(
+        new DealNotFoundError(dealId.toString()),
+      );
+    });
+
+    it('calls updateOne with the expected parameters', async () => {
+      // Act
+      await TfmDealCancellationRepo.submitDealCancellation(dealId, mockDealCancellationObject, auditDetails);
+
+      // Assert
+      const expectedFilter = {
+        _id: { $eq: new ObjectId(dealId) },
+        'tfm.stage': { $ne: TFM_DEAL_STAGE.CANCELLED },
+        'dealSnapshot.submissionType': { $in: [DEAL_SUBMISSION_TYPE.AIN, DEAL_SUBMISSION_TYPE.MIN] },
+        'tfm.cancellation.reason': { $eq: mockReason },
+        'tfm.cancellation.bankRequestDate': { $eq: mockBankRequestDate },
+        'tfm.cancellation.effectiveFrom': { $eq: mockEffectiveFrom },
+      };
+      const expectedUpdate = flatten({
+        'tfm.stage': TFM_DEAL_STAGE.CANCELLED,
+        'tfm.cancellation.status': TFM_DEAL_CANCELLATION_STATUS.COMPLETED,
+        auditRecord: generateAuditDatabaseRecordFromAuditDetails(auditDetails),
+      });
+
+      expect(getCollectionMock).toHaveBeenCalledWith(MONGO_DB_COLLECTIONS.TFM_DEALS);
+      expect(updateOneMock).toHaveBeenCalledWith(expectedFilter, expectedUpdate);
+    });
```

Comment:
Can we add `.toHaveBeenCalledTimes` assertions please

Reply:
Yep, good spot, have fixed